---
title: "countland: R tutorial"
output: rmarkdown::github_document
vignette: >
  %\VignetteIndexEntry{countland: R tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### Samuel H. Church

url: https://github.com/shchurch/countland/blob/main/countland-R/vignettes/vignette-tutorial.Rmd

This tutorial demonstrates the major functions of `countland` by applying them to a Gold standard single-cell RNA sequencing dataset from [Freytag _et al_ (2018)](10.12688/f1000research.15809.2).

`countland` is a **barebones** set of functions for applying a restricted linear algebra to the analysis of count-based data. As such, there are many opportunities for further optimization that may prove useful in the anlaysis of your own data. We provide the source code freely available at https://github.com/shchurch/countland and encourage users and developers to fork the code for their own purposes.

The following packages are required to complete the tutorial

```{r setup}
library(countland)
library(Seurat) # for reading in the data
library(ggplot2)
theme_set(theme_classic())
set.seed(84095) # set random seed for reproducibility
```

## Get the data

We have used the Gold standard dataset provided by [Freytag _et al_ (2018)](10.12688/f1000research.15809.2). This data consists of ~1000 cells that have ground truth labels corresponding to three human lung cell lines.

`countland` accepts an sparse data matrix (as does `Seurat`). 

```{r data}
gold_path <- system.file("extdata", "Gold_Freytag2018/", package = "countland", mustWork = TRUE)
gold.data <- Seurat::Read10X(data.dir = gold_path)
m <- gold.data
```

## Initialize `countland` object

Initialize `countland` by calling the core function on sparse data matrix.

```{r init}
C <- countland(m)
```

The count matrix is stored in `C@counts`

```{r count-matrix}
C@counts[1:10,1:10]
```

Note that most counts are zero for scRNA-seq data.

## Summarize counts across cells and genes

We can explore our data by checking the total number of counts and other expression measures across genes. These measures include:

* The number of cells where each gene was detected.
* The maximum observed count value per gene.
* The total number of counts per gene, summed across cells.
* The number of cells where a gene was detected at a count value larger than 1 or 10.
* The number of unique count values (e.g. 0, 1, 2). Given the discrete nature of low-magnitude count values, this measure can provide insight into expression variability across cells.
* The largest number $n$, where there are $n$ cells with $\geq n$ counts for the gene in question. We refer to this measure as the _count index_, and it can be helpful for finding genes that frequently show higher count values, as compared to genes that are mostly detected at values of 1 or 2, with a few high-count exceptions.


```{r gene-score}
# calculate expression scores across genes
C <- ScoreGenes(C)
head(C@gene_scores)
```

We can also calculate the same measures, but across cells. 

It can be helpful to see how many counts are derived from certain genes, such as mitochondrial genes. Here we chose a gene at random, its counts are listed under `feature_match_counts`. Regular expressions are permitted for name matching.

```{r cell-score}
# sum counts from genes matching this regex expression
gene_string <- "162396$"

# calculate expression scores across cells
C <- ScoreCells(C,gene_string)
head(C@cell_scores)
```

## Cluster cells by similarity

In our count-based approach, we compare cell transcriptomes by calculating the dot product of untransformed counts between all pairwise combinations of cells. This process generates a similarity matrix (rather than a distance matrix, as typically used in scRNA workflows).

```{r dots}
C <- Dot(C)
# preview the resulting similarity matrix
C@dots[1:5,1:5]
```
   
Cell similarity can then be analyzed by applying spectral embedding to this dot product similarity matrix. Spectral embedding involves calculating the graph Laplacian of the matrix and then estimating the eigenvectors and eigenvalues of this graph. 

```{r embed}
C <- Embed(C)
```

This matrix can also be used to cluster cells by applying a k-means algorithm to the eigenvector matrix (spectral clustering). In order to determine the optimal number of clusters, it can be helpful to apply an eigengap heuristic. This can help decide the optimal number of clusters, but it is only a guideline. According to this heuristic, the optimal number of clusters is _k_ where the difference in eigenvalues $| e_{k+1} - e_{k} |$ is largest.

```{r eigengap,fig.width=4,fig.height=3}
PlotEigengap(C)
```

Here we see big jumps between 1 and 2, and a smaller gap between 2 and 3. For many datasets, you may want to consider other factors, e.g. choosing a minimum number of clusters, whether or not the eigengap reflects this.

Here we have chosen 3 as the optimal number of clusters, given our prior knowledge about this benchmark dataset.

```{r cluster}
C <- Cluster(C,n_clusters=3)
```

The clustering results are stored in `C@clusters`. The cell similarity matrix can be visualized in two dimensions by plotting points using two eigenvectors.

```{r plot-cluster,fig.width=6}
PlotEmbedding(C)
```

As we see, cells fall into three clusters which separate along the first two dimensions.

## Subsampling data

It's not always necessary to standardize sequencing depth across cells in order to make useful comparisons. However there are certain comparisons where we might expect differences in sequencing depth to obscure biological differences, for example, when calculating differential gene expression across cells. 

One method of standardizing sequencing depth is to subsample counts to an equivalent number for all cells. Since the scRNA-seq counting process stops at an arbitrary point, we can randomly subsample from our observations per cell to stop the process at a specific number of our choosing.

Here we do this with the function `Subsample()`. To subsample to the minimum number of total counts across cells, use `cell_counts='min'`, otherwise use a number of counts larger than the minimum value.

```{r subsample-cells}
C <- Subsample(C,cell_counts='min')
# preview the resulting subsampled count matrix
C@subsample[1:10,1:10]
```

There are also cases where we might expect differences in the magnitude of expression across genes to obscure biological variation. The dot product is calculated by summing the product of each gene, meaning genes with substantially larger count values will contribute more to the dot product than genes with smaller values. This feature of the dot product can be useful for emphasizing genes with the largest dynamic range of counts, given that expression variance scales with magnitude. But when highly expressed genes do not contain informative biological signal, this variance might drown out signal from genes with lower expression magnitudes.

A count-based approach for tuning down the signal of highly expressed genes is to limit their total counts by randomly subsampling observations per gene. We can accomplish this by establishing a threshold for total counts and subsampling that number of observations from any gene vector that exceeds that threshold.

Here we do this with the function `Subsample(gene_counts=[maximum])`. An example maximum value might be equal to 10x the number of cells (columns in the count matrix).

```{r subsample-genes}
C <- Subsample(C,gene_counts=10*ncol(C@counts)) # doing so will overwrite our previous subsampled matrix 

# To subsample both genes and cells, use gene_counts and cell_counts in the same function. 
# Genes will be subsampled first.

#C <- Subsample(C,gene_counts=10*ncol(C@counts),cell_counts='min')
```

## Subsetting data

An alternative approach to subsampling is to focus on only subsets of genes or cells when making comparisons. Here we filter the count matrix to only certain cells and genes using `SubsetCells()` and `SubsetGenes()`.

With `countland`, such data filtering may not be necessary or helpful, so we don't recommend any filtering scheme by default.

```{r filt-cell}
# cells with fewer than 8,500 unique features
filter_cell_names <- C@cell_scores[C@cell_scores$n_features < 8500,]$names
filter_cell_index <- which(C@names_cells %in% filter_cell_names) 
C <- SubsetCells(C,filter_cell_index,remove_empty=FALSE)
```

```{r filt-gene}
# cells with greater than 100 observations
filter_gene_names <- C@gene_scores[C@gene_scores$n_cells > 100,]$names
filter_gene_index <- which(C@names_genes %in% filter_gene_names) 
C <- SubsetGenes(C,filter_gene_index,remove_empty=FALSE)
```

The original count matrix can be restored at any time.

```{r restore}
C <- RestoreCounts(C)
```

## Identify marker genes

The ideal marker gene for a cluster of cells can be definedas the gene that is most specifically expressed in cluster cells, or as the gene with the highest differential expression between cells in the cluster and all other cells.

A count-based approach for identifying marker genes by specificity is to count the number of cells with non-zero observations for each gene, and then calculating the difference between the fraction of these cells in a cluster versus the fraction that are not (here with `method='prop-zero'`). The ideal marker gene would return a value of one, indicating it was expressed in all cluster cells and no others.

```{r marker-zero}
C <- RankMarkerGenes(C,method='prop-zero',subsample=F)
C@marker_genes[(C@marker_genes$cluster_label == 1),]
```

```{r plot-marker-zero,fig.width=12}
cluster_marker <- C@marker_genes[C@marker_genes$cluster_label == 1,]
cluster_top <- cluster_marker[cluster_marker$rank == 1,]$names
gene_index = which(C@names_genes == cluster_top)
PlotMarker(C,gene_index)
```

This plot shows the counts for the top marker gene, defined by specificity, for cluster 1.

Alternatively we can defined the top marker gene by differential expression (here with `method='rank-sums'`). This is assessed using rank-sums tests. Counts for a given gene are ranked between cluster and non-cluster cells, the ranks for each group are summed, and a test statistic is calculated. This statistic is used to test the hypothesis that observations from cluster cells are larger than those from non-cluster cells.

Because heterogeneity in count depth can influence the magnitude of expression for individual gene, subsampling to a standard sequencing depth prior to calculating differential gene expression is recommended.

```{r marker-ranks}
C <- Subsample(C,cell_counts='min')
C <- RankMarkerGenes(C,method='rank-sums',subsample=T)
C@marker_genes[(C@marker_genes$cluster_label == 1),]
```

```{r plot-marker-ranks,fig.width=12}
cluster_marker <- C@marker_genes[C@marker_genes$cluster_label == 1,]
cluster_top <- cluster_marker[cluster_marker$rank == 1,]$names
gene_index = which(C@names_genes == cluster_top)
PlotMarker(C,gene_index)
```

This plot shows the counts for the top marker gene, defined by differential gene expression, for cluster 1.

